### ISSUE TITLE
"Fill match arms" code action creates non-exhaustive match

### ISSUE THREAD

**Issue by hniksic** [rust-analyzer/rust-analyzer#8129]
I have code like this:

```rust
        let mut chars = s.chars().peekable();
        match chars.peek() {
            Some(&'+') => {
                chars.next();
            }
            Some(&'-') => {
                bigint.negative = true;
                chars.next();
            }
            _ => (),
        }
```

Rust analyzer (run through Emacs) suggests the "fill match arms" action which, when applied, changes the fallback match arm to:

```rust
        None => {}
```

However, that's incorrect because as the compiler points out, the patterns `Some(&'\u{0}'..='*')`, `Some(&',')`, `Some(&'.'..='\u{d7ff}')` and 1 more not covered.

**Comment by lnicola** [comment]
Just to be clear, what do you expect "Fill match arms" to do on that code?

What about this one?

```rust
        let mut chars = s.chars().peekable();
        match chars.peek() {
            Some(&'+') => {
                chars.next();
            }
            Some(&'-') => {
                bigint.negative = true;
                chars.next();
            }
        }
```

**Comment by hniksic** [comment]
> Just to be clear, what do you expect "Fill match arms" to do on that code?

To be honest, I don't expect a "fill match arms" action to be suggested for that code. I clicked the suggested action because I was curious what the analyzer would do, and it actually created invalid code, which seemed like a bug.

I am a new user of rust-analyzer, so maybe I'm misunderstanding the intention behind the suggestions (code actions).

> What about this one? [...]

I guess adding a `_` fallback might be ok for this case. But in my view it would also be fine to just not offer the suggestion for cases where the analysis can't be done or where listing all options would be infeasible for other reasons.

**Comment by lnicola** [comment]
Yeah, I think we shouldn't offer it in cases like these. Just wanted to make sure you didn't want us to list every `char`.

**Comment by flodiebold** [comment]
I think adding `Some(_)` and `None` would be correct.

**Comment by hniksic** [comment]
@lnicola I didn't want the analyzer to list every `char`. :)

@flodiebold Adding `Some(_)` and `None` seem correct for the case in Inicola's comment, but less so for my original case. The `_` fallback was already there, so a suggestion that additionally distinguishes between `Some(_)` and `None` doesn't seem to add much value.

**Comment by flodiebold** [comment]
No, I think it's still the right thing in the original case. It's no different from when you have
```rust
match option {
    _ => {}
}
```
and apply the assist.

**Comment by hniksic** [comment]
Fair enough - I guess the issue is that I misinterpreted the assist as a sort of suggestion.

**Comment by flodiebold** [comment]
It's only a suggestion if it comes with a diagnostic, otherwise it's just something to help transform the code if you want to. We have many cases of assists doing opposite transformations :slightly_smiling_face:
